CMake cores: fix cpu/fpu flags

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Couple of cores did not have common flag `--mcpu` set. I also aligned fpu.

This one goes along with #14190  - both are needed to have both toolchains fixed for couple of cores.

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [x] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

0xc0170 committed:
CMake cores: fix cpu flag

Couple of cores did not have common flag `--mcpu` set

kjbracey commented on tools/cmake/cores/Cortex-A9.cmake:
```diff
@@ -9,6 +9,7 @@ if(${MBED_TOOLCHAIN} STREQUAL "GCC_ARM")
         "-mfpu=vfpv3"
         "-mfloat-abi=hard"
         "-mno-unaligned-access"
+        ""-mcpu=cortex-a9""
```
Excessive quoting. Shouldn't this replace the `-march=armv7-a`?

0xc0170 replied:
Good point, checked https://community.arm.com/developer/tools-software/tools/b/tools-software-ides-blog/posts/compiler-flags-across-architectures-march-mtune-and-mcpu 

`On Arm, -mcpu combines -march and -mtune.4 On x86, -mcpu and -mtune are the same.5` .

`-mcpu=X: On Arm, this flag is a combination of -march and -mtune.  It simultaneously specifies the target architecture and optimizes for a given microarchitecture.  On x86 this flag is a deprecated synonym for -mtune.`

I'll remove it

0xc0170 replied:
This is for Gcc: 
`This specifies the name of the target ARM processor. GCC uses this name to derive the name of the target ARM architecture`

kjbracey commented on tools/cmake/cores/Cortex-M33F.cmake:
```diff
@@ -8,6 +8,7 @@ if(${MBED_TOOLCHAIN} STREQUAL "GCC_ARM")
         "-mfpu=fpv5-sp-d16"
         "-mfloat-abi=softfp"
         "-march=armv8-m.main"
+        "-mcpu=cortex-m33+nodsp"
```
Again, this should be replacing the `-march`. That would possibly have been used due to older toolchains not knowing "cortex-m33". I assume that's valid with whatever we require now?

0xc0170 replied:
fixed now (I also fixed dsp options as there were again errors between march or toolchains).

0xc0170 committed:
CMake cores: remove march, not needed if we use mcpu

0xc0170 committed:
CMake cores: fix dsp for M33FE (enabling it)

I realized we specified nodsp for M33 core but had march for FE parts enabled with dsp.
This fixes it.

M33F has no dsp. M33FE cores have dsp enabled.

kjbracey commented on tools/cmake/cores/Cortex-M33FE-NS.cmake:
```diff
@@ -7,11 +7,11 @@ if(${MBED_TOOLCHAIN} STREQUAL "GCC_ARM")
         "-mfpu=fpv5-sp-d16"
         "-mfloat-abi=softfp"
-        "-march=armv8-m.main+dsp"
+        "-mcpu=cortex-m33+dsp"
```
Is `+dsp` valid here? Maybe, but don't see it in the docs. They just say `cortex-m33` DSP is assumed, and you can specify `+nodsp`. (Note we're also not specifying `+fp`).

https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html

`-mcpu` works differently to `-march` in this regard - you're always enabling extensions in `-march`, but usually disabling them in `-mcpu`.

0xc0170 replied:
I'll remove dsp from this one

kjbracey commented on tools/cmake/cores/Cortex-M33FE-NS.cmake:
```diff
@@ -7,11 +7,11 @@ if(${MBED_TOOLCHAIN} STREQUAL "GCC_ARM")
 elseif(${MBED_TOOLCHAIN} STREQUAL "ARM")
     list(APPEND common_options
-        "-mcpu=cortex-m33"
+        "-mcpu=cortex-m33+dsp"
```
Unlike for GCC, this is perfectly fine according to the docs, as `+X` and `+noX` are always available. But if you do want to say `+dsp` to be explicit, why not `+fp` too?

I would imagine ARMC6's defaults are the same as GCC's so might be best to keep it the same as GCC.

0xc0170 replied:
fixed, I only followed for now what we had (separate fpu settings for both toolchains). To keep both toolchains the same.

kjbracey replied:
Yeah, it's not clear from docs how `+fp`/`+nofp` (or lack of and hence default) and `-mfpu` interact. Maybe don't mess with whatever we know/think works from original python stuff.

0xc0170 committed:
CMake cores: remove dsp as it is on by default

0xc0170 committed:
CMake cores: align fpu settings

Move the fpu settings across all cores so they are aligned.
Some cores did not have fpu set for ARMClang. I matched here both toolchains to set up
the fpu the same (Only exception is soft vs hard - that is separate issue I would not like to touch
in this series).

jeromecoutant reviewed (approved):
I could verify with blinky application on:
- NUCLEO_L552ZE_Q ("Cortex-M33FE")
- NUCLEO_F767ZI ("Cortex-M7F")
- NUCLEO_H743ZI2 ("Cortex-M7FD")
Build and execution OK with ARM and GCC
